_posts/2019-06-05-uncertainty-estimates-tfprobability/uncertainty_tfprobability.Rmd

```diff
@@ -0,0 +1,779 @@
+
+Finally, the notions of _aleatoric_ and _epistemic_ uncertainty, which may have stayed a bit abstract in the prior post, should get much more concrete here.
+
+## Aleatoric vs. epistemic uncertainty
+
+Reminiscent somehow of the classic decomposition of generalization error into bias and variance, splitting uncertainty into its epistemic and aleatoric constituents separates an irreducible from a reducible part.
+
+The reducible part relates to imperfection in the model: In theory, if our model were perfect, epistemic uncertainty would vanish. Put differently, if the training data were unlimited -- or if they comprised the whole population -- we could just add capacity to the model until we've obtained a perfect fit.
+
+In contrast, normally there is variation in our measurements. There may be one true process that determines my resting heart rate; nonetheless, actual measurements will vary over time. There is nothing to be done about this: This is the aleatoric part that just remains, to be factored into our expectations.
+
+Now reading this, you might be thinking: "Wouldn't a model that actually _were_ perfect capture those pseudo-random fluctuations?". We'll leave that phisosophical question be;  instead, we'll try to illustrate the usefulness of this distinction by example, in a practical way. In a nutshell, viewing a model's _aleatoric_ uncertainty output should caution us to factor in appropriate deviations when making our predictions, while inspecting _epistemic_ uncertainty should help us re-think the appropriateness of the chosen model.
+
+Now let's dive in and see how we may accomplish our goal with `tfprobability`. We start with the simulated dataset.
+
+## Uncertainty estimates on simulated data
+
+### Dataset
+
+
+We re-use the dataset from the Google TensorFlow Probability team's [blog post on the same subject]( https://medium.com/tensorflow/regression-with-probabilistic-layers-in-tensorflow-probability-e46ff5d37baf) ^[see also the corresponding [notebook](https://colab.research.google.com/github/tensorflow/probability/blob/master/tensorflow_probability/examples/jupyter_notebooks/Probabilistic_Layers_Regression.ipynb#scrollTo=5zCEYpzu7bDX)], with one exception: We extend the range of the independent variable a bit on the negative side, to better demonstrate the different methods' behaviors.
+
+Here is the data-generating process. We also get library loading out of the way. Like in the preceding posts on `tfprobability`, please use the development versions of `tensorflow` and `tfprobability` as well as `keras`. Call `install_tensorflow(version = "nightly")` to obtain a current nightly build of TensorFlow and TensorFlow Probability:
+
+
+```{r}
+# make sure we use the development versions of tensorflow, tfprobability and keras
+devtools::install_github("rstudio/tensorflow")
+devtools::install_github("rstudio/tfprobability")
+devtools::install_github("rstudio/keras")
```

Comment:
Why do we need `keras` and `tensorflow` devel? I guess to use nightly?

Reply:
there've been quite a few changes since the CRAN release, one of which involves load order of packages (keras/tensorflow), i'm pretty sure it's safest to use nightly for all 3 here...

Reply:
If you still have the error around, I would share it to make sure there is not something "broken" in tensorflow/keras, would be nice to understand why we can't use `tfprobability` with older TensorFlow versions... or have `tf.probability` already state they are not backwards compatible and the latest version of TF is required?

Reply:
Sorry I was unclear, I guess it's 2 different things:

1. we need to push some changes (that loading order one especially) to CRAN for seemless integration of `keras` and `tfprobability` (the R packages)

2. TFP (Python) always needs a corresponding build of TF (so, nightly needs nightly). And the functionality I use is not the - pretty old! - TFP release 0.6.

Reply:
Gotcha! So the post itself requires nightly TFP. So I would change this:

> Like in the preceding posts on tfprobability, please use the development versions of tensorflow and tfprobability as well as keras.

to

> Like in the preceding posts on tfprobability, this post also makes use of new tfprobability features, so please use the development versions of tensorflow and tfprobability as well as keras.

Reply:
done!